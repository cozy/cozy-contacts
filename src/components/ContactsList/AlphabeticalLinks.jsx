import React from 'react'
import PropTypes from 'prop-types'

import Button from 'cozy-ui/transpiled/react/Button'

const filterContactInitials = categorizedContacts => {
  delete categorizedContacts.EMPTY
  return Object.keys(categorizedContacts).sort()
}

export const AlphabeticalLinks = ({ categorizedContacts, onClick }) => {
  const contactsInitials = filterContactInitials(categorizedContacts)

  return (
    <div>
      {contactsInitials.map(letter => (
        <Button
          subtle
          key={letter}
          label={letter.toUpperCase()}
          onClick={() => onClick(letter)}
          className="alphabetical-link"
        />
      ))}
    </div>
  )
}

AlphabeticalLinks.propTypes = {
  categorizedContacts: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired
}

export default AlphabeticalLinks
