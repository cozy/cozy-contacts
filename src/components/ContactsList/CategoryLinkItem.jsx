import React from 'react'
import PropTypes from 'prop-types'

import Chips from 'cozy-ui/transpiled/react/Chips'

const CategoryLinkItem = ({ category }) => {
  const scrollTo = e => {
    e.preventDefault() // Stop Page Reloading
    let categoryElement = document.getElementById(`cat-${category}`)
    categoryElement &&
      categoryElement.scrollIntoView({ behavior: 'smooth', block: 'start' })
  }

  return (
    <Chips
      sx={{ margin: '10px' }}
      label={category.toUpperCase()}
      onClick={scrollTo}
    />
  )
}

CategoryLinkItem.propTypes = {
  category: PropTypes.string.isRequired
}

export default React.memo(CategoryLinkItem)
