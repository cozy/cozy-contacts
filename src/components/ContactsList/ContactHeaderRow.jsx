import React from 'react'
import PropTypes from 'prop-types'

const ContactHeaderRow = props => <div className="divider">{props.header}</div>

ContactHeaderRow.propTypes = {
  header: PropTypes.string.isRequired
}
export default ContactHeaderRow
