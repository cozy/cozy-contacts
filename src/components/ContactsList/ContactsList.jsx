import React from "react";
import PropTypes from "prop-types";
import { sortGivenNameFirst, getDisplayedName } from "./";
import ContactRow from "./ContactRow";

const ContactHeaderRow = props => (
  <div className="contact-header">{props.header}</div>
);
ContactHeaderRow.propTypes = {
  header: PropTypes.string.isRequired
};

const ContactsEmptyList = () => (
  <div>
    <p>No Contact</p>
    <p>Try to create one with the big button on the top</p>
    <p>
      Or to import your contacts with this <a href="#">link</a>
    </p>
  </div>
);

const ContactsList = props => {
  if (props.contacts.length === 0) {
    return <ContactsEmptyList />;
  }
  const sortedContacts = [...props.contacts].sort(sortGivenNameFirst);
  const rows = [];
  let lastLetter = null;
  sortedContacts.forEach(contact => {
    const name = getDisplayedName(contact);
    if (name[0] !== lastLetter) {
      rows.push(<ContactHeaderRow header={name[0] || "EMPTY"} />);
    }
    rows.push(
      <ContactRow
        key={contact._id}
        contact={contact}
        selection={
          props.onSelect && {
            onSelect: () => {
              props.onSelect(contact._id);
            },
            selected: props.selection.includes(contact._id)
          }
        }
      />
    );
    lastLetter = name[0];
  });
  return <div>{rows}</div>;
};
ContactsList.propTypes = {
  contacts: PropTypes.array.isRequired,
  onSelect: PropTypes.func,
  selection: PropTypes.array
};
ContactsList.defaultProps = {
  onSelect: null,
  selection: []
};

export default ContactsList;
