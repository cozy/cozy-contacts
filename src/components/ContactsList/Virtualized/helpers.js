export const makeColumns = ({ t, isMobile }) => {
  const mobileColumns = [
    {
      id: 'fullname',
      label: t('fields.familyName'),
      width: '100%',
      sortable: false
    },
    {
      id: 'actions',
      width: 40,
      label: '',
      textAlign: 'center',
      disableClick: true,
      sortable: false
    }
  ]

  if (isMobile) {
    return mobileColumns
  }

  return [
    {
      id: 'fullname',
      disablePadding: true,
      label: t('fields.familyName'),
      width: '45%',
      sortable: false
    },
    {
      id: 'email.[0].address',
      label: t('fields.email'),
      textAlign: 'left',
      width: '30%',
      sortable: false
    },
    {
      id: 'phone.[0].number',
      label: t('fields.phone'),
      textAlign: 'left',
      width: 'auto',
      sortable: false
    },
    {
      id: 'actions',
      width: 40,
      label: '',
      textAlign: 'center',
      disableClick: true,
      sortable: false
    }
  ]
}
