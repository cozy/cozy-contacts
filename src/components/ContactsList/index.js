import ContactsList from './ContactsList'

export function buildLastNameFirst(contact) {
  const givenNameA =
    contact.name && contact.name.givenName
      ? contact.name.givenName.toUpperCase()
      : ''
  const familyNameA =
    contact.name && contact.name.familyName
      ? contact.name.familyName.toUpperCase()
      : ''
  const nameA = `${familyNameA} ${givenNameA}`.trim()
  return nameA
}

export const sortLastNameFirst = (contact, comparedContact) => {
  const nameA = buildLastNameFirst(contact)
  const nameB = buildLastNameFirst(comparedContact)
  return nameA.localeCompare(nameB)
}

export default ContactsList
