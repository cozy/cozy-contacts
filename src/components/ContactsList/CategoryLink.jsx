import React from 'react'
import PropTypes from 'prop-types'
import CategoryLinkItem from './CategoryLinkItem'

const CategoryLink = ({ categories }) => {
  return (
    <div className="category-link">
      {categories.map(category => (
        <CategoryLinkItem key={`link-cat-${category}`} category={category} />
      ))}
    </div>
  )
}

CategoryLink.propTypes = {
  categories: PropTypes.array.isRequired
}

export default React.memo(CategoryLink)
