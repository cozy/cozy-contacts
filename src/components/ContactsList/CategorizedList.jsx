import React from 'react'
import PropTypes from 'prop-types'

import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import ContactHeaderRow from './ContactHeaderRow'
import ContactsSubList from './ContactsSubList'
import { categorizeContacts } from '../../helpers/contactList'

const CategorizedList = ({ contacts }) => {
  const { t } = useI18n()
  const categorizedContacts = categorizeContacts(contacts, t('empty-list'))

  return (
    <ol className="list-contact">
      {Object.entries(categorizedContacts).map(([header, contacts]) => (
        <li key={`cat-${header}`}>
          <ContactHeaderRow key={header} header={header} />
          <ContactsSubList contacts={contacts} />
        </li>
      ))}
    </ol>
  )
}

CategorizedList.propTypes = {
  contacts: PropTypes.array.isRequired
}

export default CategorizedList
