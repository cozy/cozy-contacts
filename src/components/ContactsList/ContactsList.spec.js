import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import AppLike from '../../tests/Applike'
import ContactsList from './ContactsList'

import { johnDoeContact } from '../../helpers/testData'

const fakeContactWithIndexes = {
  ...johnDoeContact,
  indexes: {
    byFamilyNameGivenNameEmailCozyUrl: 'john.doe@cozycloud.cc'
  }
}

describe('ContactsList', () => {
  it('should match the snapshot', () => {
    const tree = renderer
      .create(
        <AppLike>
          <ContactsList contacts={[fakeContactWithIndexes]} />
        </AppLike>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should renders an empty state', () => {
    const contactsListInstance = (
      <AppLike>
        <ContactsList contacts={[]} />
      </AppLike>
    )
    const contactslist = mount(contactsListInstance)
    const contactsemptylist = contactslist.find('ContactsEmptyList')
    expect(contactsemptylist).toBeDefined()
  })
})
