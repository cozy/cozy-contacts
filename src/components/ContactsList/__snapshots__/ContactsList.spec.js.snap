// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsList should match the snapshot 1`] = `
<div
  className="list-wrapper"
>
  <ol
    className="list-contact"
  >
    <li
      id="contact-group-j"
    >
      <div
        className="divider"
      >
        j
      </div>
      <ol
        className="sublist-contact"
      >
        <li>
          <div
            className="contact"
            onClick={[Function]}
          >
            <div
              className="contact-selection"
              onClick={[Function]}
            >
              <span
                data-input="checkbox"
              >
                <input
                  checked={false}
                  readOnly={true}
                  type="checkbox"
                />
                <label />
              </span>
            </div>
            <div
              className="contact-identity"
            >
              <div
                className="styles__c-avatar___PpDI- styles__c-avatar--small___1Y_Pv"
                style={
                  Object {
                    "backgroundColor": "var(--brightSun)",
                    "color": "white",
                  }
                }
              >
                
                <span
                  className="styles__c-avatar-initials___310qC"
                >
                  JD
                </span>
              </div>
              <div
                className="u-ellipsis u-ml-1"
              >
                <span
                  className=""
                >
                  John
                   
                </span>
                <span
                  className="u-fw-bold"
                >
                  Doe
                   
                </span>
              </div>
            </div>
            <div
              className="contact-email u-ellipsis"
            >
              john.doe@cozycloud.cc
            </div>
            <div
              className="contact-phone u-ellipsis"
            >
              +33 (2)0 90 00 54 04
            </div>
            <div
              className="contact-cozyurl u-ellipsis"
            >
              https://johndoe.mycozy.cloud
            </div>
          </div>
        </li>
      </ol>
    </li>
  </ol>
  <div />
</div>
`;
